### DISCUSSION TITLE
Redefine something in the "FreeRTOSConfig.h" on a user sketch

### DISCUSSION THREAD

**Discussion by JoshuaKim6502** [feilipu/Arduino_FreeRTOS_Library#55] [Q&A]
Hi Phillip,
I enjoy using the FreeRTOS on the Arduino Uno.  I understand why the header file "FreeRTOSConfig.h" is not in an individual project but the Ardunio/libraries/FreeRTOS/src.

What if I put a needed line to my source sketch, while the FreeRTOSConfig.h is intact? I think it works. Correct?

For example, I copy this line to my sketch;

>#undef  configMAX_PRIORITIES
>#define configMAX_PRIORITIES    (1+5)

while the FreeRTOSConfig.h has the original;

>#define configMAX_PRIORITIES    3

It looks work when I compile with no warning message.

Joshua

**Comment by feilipu** [comment]
Joshua,
I'm not sure what will happen.
I suggest that you assign a variable in your sketch with `configMAX_PRIORITIES`, and then print it.
First do it without redefining in your sketch, to ensure you're seeing the original number `3`, and then again with your redefinition to ensure you see the new number `(1+5)`.
Best regards, Phillip

**Comment by JoshuaKim6502** [comment]
Thanks for the comment.  I tested again as you recommend.
Without redefining the `configMAX_PRIORITIES` value, my code prints the original **`3`** as expected.
Once I redefine the value to `6` in my source, the result printed is **`6 = (1+5)`**.
So the redefinition on the user source looks work.

Thanks again, Joshua

**Comment by JoshuaKim6502** [comment]
I misunderstood it. After precise diagnostic concerning the `configMAX_PRIORITIES` value, I realized that **redefining it on the user source is meaningless** even though it looks to be changed.
The original `<FreeRTOSConfig.h>` dominates compiling kernel codes first off, while any redefined on the user source appears last.

**Reply by Mathullah** [reply]
I am facing the same problem. Have you found a solution in the meantime?

**Reply by feilipu** [reply]
The best option is to change it inside the library, as it is source code and is compiled each build. Just look for the library/FreeRTOS folder or directory in your Arduino installation.

**Reply by Mathullah** [reply]
That's ok as long as the library is used just for a single project. It gets very awkward when you're working on multiple projects on different machines.
I'm wondering whether its possible to move the corresponding files from the library to the project and manage only the templates in the library? Then, for example, it would be easier to set up a project-specific tick routine.

**Reply by feilipu** [reply]
TBH, it sounds like you've graduated from using this library. Congratulations.

It is really just meant to be a simple yet complete introduction to FreeRTOS.

You should probably move the library files into your own source control and IDE. That way you can have fine control over what changes, and when.

As much as I enjoy the idea of using this library for production, I'm afraid my testing is too lax to ensure that there are no subtle issues lurking.

Arduino actively masks complexity for (new) users. That's the deal you get.

**Reply by JoshuaKim6502** [reply]
Hello Mathullah, thanks for your interest in the issue.
I compromise to edit the master FreeRTOSConfig.h for each project while copying it to the individual work folder for reference only.
At first, I could modify the library or others to retain an independent FreeRTOSConfig.h file for each project. But for me, it was to be easier said than done.  The Arduino is not my professional working platform but an auxiliary to reassure an idea or algorithm. So I went for an easy way with myself. 
If you make a better way, please show it off.

**Reply by Mathullah** [reply]
Chaps,
Yes I'm quite aware that this library is not intended for professional use and I agree that the focus should be on a simple usage. Nevertheless, it would be nice if project specific settings were possible.
I use multiple host machines depending on where I work. This depends mostly on my wife since she has priority by choosing the right machine (for watching some netflix stuff on the 16" macbook in the bed🤷‍♂️). In this case I have to go back to our primal mac...
Another aspect is that I work together (not a commercial project) with someone who's not familiar with embedded sw development. That's why I went with Arduino. Otherwise I would probably use something more capable.
@feilipu I will try your approach by coppying the library code into the project. Hope this works for me. Had some problems in the past with sub folders for code. Arduino could not find / compile them.
